### PR TITLE
Add a red-black tree implementation and testing.

### PR DESCRIPTION
**Pull request by mjp41**

### CODE REVIEW

**Review by nwf — COMMENTED** [review]
LGTM save for some nits (mostly commentary and naming) and a request for a `concept`.  I have not attempted to verify correctness of all the cases.

FWIW, you might like reading https://www.cs.princeton.edu/~rs/talks/LLRB/LLRB.pdf if you haven't before.

File: src/ds/redblacktree.h [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+      if (Rep::is_red(curr))
+        return left_inv;
+      else
+        return left_inv + 1;
+    }
+
+    struct RBStep
+    {
+      ChildRef node;
+      bool dir = false;
```
[the remark this answers is not in the document]
Comment (mjp41) [reply src/ds/redblacktree.h:189]: I am not sure about this change.  It is nice to have !dir just work without having to write stuff.

File: src/ds/redblacktree.h [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+      path.curr() = value;
+      get_dir(true, path.curr()) = Rep::null;
+      get_dir(false, path.curr()) = Rep::null;
+      Rep::set_red(value, true);
+
+      debug_log("Insert ", path);
+
+      // Propogate double red up to rebalance.
+      // These notes were particularly clear for explaining insert
+      // https://www.cs.cmu.edu/~fp/courses/15122-f10/lectures/17-rbtrees.pdf
```
[the remark this answers is not in the document]
Comment (mjp41) [reply src/ds/redblacktree.h:572]: He had some nice notes, that nicely captured the symmetry.

File: src/ds/redblacktree.h [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+      // rebalance. The following slides nicely cover the case analysis below
+      //   https://www.cs.purdue.edu/homes/ayg/CS251/slides/chap13c.pdf
+      while (path.curr() != ChildRef(root))
+      {
+        key parent = path.parent();
+        bool cur_dir = path.curr_dir();
+        key sibling = get_dir(!cur_dir, parent);
+
+        // Handle red sibling case.
+        // This performs a rotation to give a black sibling.
```
[the remark this answers is not in the document]
Comment (mjp41) [reply src/ds/redblacktree.h]: Yep. I'll add a comment

File: src/ds/redblacktree.h [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+      ChildRef(H& p) : ptr(&p) {}
+
+      operator K()
+      {
+        return Rep::get(ptr);
+      }
+
+      K operator=(K t)
+      {
+        Rep::set(ptr, t); // Only modify correct bits
```
[the remark this answers is not in the document]
Comment (mjp41) [reply src/ds/redblacktree.h]: It has to potentially preserve more.  I have adjusted the comment.

File: src/ds/redblacktree.h [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+  template<typename Rep, bool run_checks = true, bool TRACE = false>
+#endif
+  class RBTree
+  {
+    // Container that behaves like a C++ Ref type to enable assignment
+    // to treat left, right and root uniformly.
+    class ChildRef
+    {
+      using H = typename Rep::Holder;
+      using K = typename Rep::Contents;
```
[the remark this answers is not in the document]
Comment (mjp41) [reply src/ds/redblacktree.h]: Good idea.

File: src/ds/redblacktree.h [hunk 6]
```diff
@@ -0,0 +1,633 @@
...
+   * The template parameter Rep provides the representation of the nodes as a
+   * collection of functions and types that are requires.  See the associated
+   * test for an example.
+   *
+   * run_checks enables invariant checking on the tree. Enabled in Debug.
+   * TRACE prints all the sets of the rebalancing operations. Only enabled by
+   * the test when debugging a specific failure.
+   */
+#if defined(NDEBUG)
+  template<typename Rep, bool run_checks = false, bool TRACE = false>
```
[the remark this answers is not in the document]
Comment (mjp41) [reply src/ds/redblacktree.h]: Added.  Hopefully, it doesn't annoy me too much ;-)

[the remark this answers is not in the document]
Comment (mjp41) [reply src/ds/redblacktree.h]: P.s. Thanks for writing it. I find the syntax intractible, I can just about read it.

File: src/ds/redblacktree.h [hunk 7]
```diff
@@ -0,0 +1,633 @@
...
+   *
+   * The template parameter Rep provides the representation of the nodes as a
+   * collection of functions and types that are requires.  See the associated
+   * test for an example.
+   *
+   * run_checks enables invariant checking on the tree. Enabled in Debug.
+   * TRACE prints all the sets of the rebalancing operations. Only enabled by
+   * the test when debugging a specific failure.
+   */
+#if defined(NDEBUG)
```
[the remark this answers is not in the document]
Comment (mjp41) [reply src/ds/redblacktree.h]: Done.  I have split this into a separate commit, that applies this change broadly.

**Review by nwf — APPROVED** [review]
LGTM

**Pull request merged**